components/FA2Transfer.tsx

```diff
@@ -0,0 +1,271 @@
+import renderError from "./formUtils";
+
+type props = {
+  index: number;
+  remove: (index: number) => void;
+  setFieldValue: (name: string, value: any) => void;
+  getFieldProps: (name: string) => FieldInputProps<any>;
+};
+
+type fa2Token = {
+  id: number;
+  balance: string;
+  account: {
+    address: string;
+  };
+  token: {
+    id: number;
+    contract: { address: string };
+    metadata: {
+      name: string;
+      symbol: string;
+      thumbnailUri?: string;
+      displayUri?: string;
+    };
+    standard: string;
+    tokenId: string;
+    totalSupply: string;
+  };
+};
+
+type option = {
+  id: string;
+  value: string;
+  label: string;
+  tokenId: string;
+  thumbnailHash: string | undefined;
+  contractAddress: string;
+  token: fa2Token;
+};
+
+const FETCH_COUNT = 20;
+
+const tokenToOption = (fa2Token: fa2Token) => {
+  const { token } = fa2Token;
+  return {
+    id: token.id.toString(),
+    tokenId: token.tokenId,
+    value: token.id.toString(),
+    label: token.metadata.name,
+    thumbnailHash: (
+      token.metadata.thumbnailUri ??
+      token.metadata.displayUri ??
+      ""
+    ).replace("ipfs://", ""),
+    contractAddress: token.contract.address,
+    token: fa2Token,
+  };
+};
+
+const FA2Transfer = ({
+  index,
+  setFieldValue,
+  remove,
+  getFieldProps,
+}: props) => {
+  const state = useContext(AppStateContext)!;
+
+  const [isFetching, setIsFetching] = useState(true);
+  const [canSeeMore, setCanSeeMore] = useState(true);
+
+  const [filterValue, setFilterValue] = useState<string>("");
+  const [currentToken, setCurrentToken] = useState<fa2Token | undefined>();
+  const [options, setOptions] = useState<option[]>([]);
+  const fetchOffsetRef = useRef(0);
+
+  const makeName = (key: string) => `transfers.${index}.values.${key}`;
+
+  const updateValues = (newToken: fa2Token) => {
+    setCurrentToken(newToken);
+    setFieldValue(makeName("token"), newToken ?? "");
+    setFieldValue(makeName("tokenId"), newToken?.token.tokenId ?? "");
+    setFieldValue(
+      makeName("fa2Address"),
+      newToken?.token.contract.address ?? ""
+    );
+  };
+
+  const fetchTokens = useCallback(
+    (value: string, offset: number) =>
+      fetch(
+        // `${API_URL}/v1/tokens/balances?account=${state.currentContract}&offset=${offset}&limit=20&token.metadata.name.as=*${value}*`
+        `${API_URL}/v1/tokens/balances?account=tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb&offset=${offset}&limit=20&token.metadata.name.as=*${value}*`
+      )
+        .catch(e => {
+          console.log(e);
+          return {
+            json() {
+              return Promise.resolve([]);
+            },
+          };
+        })
+        .then(res => res.json())
+        .then((v: fa2Token[]) => {
+          setCanSeeMore(v.length === FETCH_COUNT);
+
+          return Promise.resolve(v);
+        }),
+    [state.currentContract]
+  );
+
+  useEffect(() => {
+    const value = getFieldProps(makeName("token")).value as fa2Token | "";
```

Comment:
Not sure how this ended here, I wanted to write `??`

Reply:
can return of `getFieldProps` be without `any`?

Reply:
as the type is `fa2Token | undefined`, it's safe to remove `| ""`.  Later on, if it's `value` is `undefined` in here ` if (!value) return;`, it will return. otherwise, it will continue.